Make tests and get the analysis to pass them

outofculture opened the pull request.

outofculture committed:
unrelated changes

outofculture committed:
bugfix: spelling

outofculture committed:
use NEURON for the model cell (still failing)

outofculture committed:
create the Vector after the IClamp or they won't communicate

outofculture committed:
no, what actually matters is holding onto references

outofculture committed:
cleanup; paramtrize

outofculture committed:
plot legend, too

outofculture committed:
new way of modeling access resistance; refactor to not use globals

outofculture committed:
explode into more tests for fine-grained look at what is failing

outofculture committed:
conic pipette sorta working?!

outofculture committed:
2 conic sections with lots of segments

outofculture committed:
manual segmentation to get equal resistance per segment

outofculture committed:
connect the segmented pipette to VC mode

outofculture committed:
more playing with the tests

outofculture committed:
access resistance was being counted in input resistance

outofculture committed:
notes and failed attempts at additional modeling

outofculture committed:
comment out "only" param now that most tests are passing

outofculture committed:
fix labels; increase granularity of pipette simulation

outofculture committed:
conditionally create soma, allowing bath simulation
 - pull expected data from args only, since soma may not exist

outofculture committed:
leaky cells have smaller resting potentials

outofculture committed:
refactor: rename, break apart mock test pulse
optional assertion call

outofculture committed:
rename model arg for clarity

outofculture committed:
first attempt at explicit hessian curve fit

outofculture committed:
BROKEN mostly hand-checked hessian

outofculture committed:
finished hang-checking the hessian

outofculture committed:
trying to tease out this broken fit

outofculture committed:
slight variation on assertion

outofculture committed:
hide warnings

outofculture committed:
does a single-pass fit work at all?

outofculture committed:
broken one-pass. refactored two-pass, trying to figure it out

outofculture committed:
dump/load interface for test pulse

outofculture committed:
shorten names to fit in smaller plots

outofculture committed:
handle loading test pulse stacks, too

outofculture committed:
no print

outofculture committed:
put the start_time in analysis for identification purposes

outofculture committed:
allow multiple H5 files to back a single stack

outofculture committed:
cache timestamp array for faster searching

outofculture committed:
fix incorrect hessian (sign, 2nd-derivative of e^(1/x))

outofculture committed:
unrelated changes

outofculture committed:
just return the two values instead of clumping them together unneccesarily

campagnola committed:
Merge branch 'master' into exp-fit-simplified

outofculture committed:
put exact_exp_fit in place (from Luke)

outofculture committed:
no need for tau_init

campagnola committed:
add more exp fitting unit tests

outofculture committed:
xoffset for the exact_fit

campagnola committed:
test fits on exact method

outofculture committed:
Merge remote-tracking branch 'origin/exp-fit-simplified' into can-we-pass

Conflicts:
	neuroanalysis/tests/test_exp_fit.py - I mostly just accepted all of origin's changes.

outofculture committed:
for exact fit: add nrm to the se in err, handle xoffset.

campagnola committed:
more test plotting

campagnola committed:
Merge remote-tracking branch 'outofculture/can-we-pass' into exp-fit-simplified

campagnola committed:
reduce size of exp tests

outofculture committed:
allow TSeries ± TSeries

outofculture committed:
this parameter is not a yoffset

outofculture committed:
current failures for detailed inspection

outofculture committed:
the transient fit is not helping pass any tests

campagnola committed:
Merge remote-tracking branch 'outofculture/can-we-pass' into exp-fit-simplified

campagnola committed:
rename test_tau

outofculture committed:
include the second fit for inspection purposes

outofculture committed:
rename variable to match its now source

outofculture committed:
ground the pipette in bath mode

outofculture committed:
bath is behaving well, but our tests were too sensitive around 0

outofculture committed:
use the default solver

outofculture committed:
comment; whitespace

outofculture committed:
parametrize exp_fit tests

outofculture committed:
allclose still fails, but the error is pretty good

outofculture committed:
add more info to test-pulse analysis plots

outofculture committed:
cleanup; live test data

outofculture committed:
quantify fit confidence with stddev of the error surface

- use that to pick either exp decay fit or simple bath fit

outofculture committed:
higher threshold; put even simple tau and y0 into fit

outofculture committed:
fix test plot

outofculture committed:
refactor: variable

outofculture committed:
use a measure of linearity as our confidence score

outofculture committed:
tighten our threshold; fix plot; make the tests easier

outofculture committed:
so noisy warnings

outofculture committed:
IC capacitance was calculated this way in ACQ4

outofculture committed:
revert! tau for a discharging capacitor only has the one resistor

outofculture committed:
import cleanup

outofculture committed:
brief notes on the use of tau

outofculture committed:
add a source

outofculture committed:
more detailed defense of our uses of tau

outofculture committed:
remove unused code

outofculture committed:
use existing save/load (but support old data for now)

outofculture committed:
no importing deleted functions

outofculture committed:
PatchClampTestPulse.__init__ does not need a stimulus arg

outofculture committed:
unused, unneeded anymore

outofculture committed:
rename to match other neuroanalysis classes

campagnola commented on neuroanalysis/fitting/exp.py:
```diff
@@ -24,8 +24,8 @@ def estimate_exp_params(data):
-    start_y = data.data[:len(data.data)//100].mean()
-    end_y = data.data[-len(data.data)//10:].mean()
+    start_y = data.data[:len(data.data)//10].mean()
+    end_y = data.data[-len(data.data)//100:].mean()
```
Why these?

outofculture replied:
the changes at the beginning of the expected curve are more drastic than the changes at the end. thus, the last hundred samples are likely going to be the same, while the first hundred samples are not.

ooooh, wait, I just read the math. it's not count of samples, it's "tenth of all samples" v. "hundredth of all samples". I read it backwards.

campagnola commented on neuroanalysis/fitting/exp.py:
```diff
@@ -42,64 +42,103 @@ def normalized_rmse(data, params, fn: Callable=exp_decay):
+    return exp_like
+
+
+def exact_fit_exp(data: TSeries):
```
Rename to exp_fit

outofculture committed:
I read the math backwards; hundredth, not hundred

outofculture committed:
only one fitter left, it gets the name

campagnola commented on neuroanalysis/test_pulse.py:
```diff
@@ -25,27 +32,72 @@ def __init__(self, rec: PatchClampRecording, indices=None):
         self.fit_trace_with_transient = None
-        self.initial_double_fit_trace = None
+
+    def save(self):
```
Seems like this wil make it awkward to add a Recording.save later on.. Could this done more cleanly by starting there instead? And also adding TSeries.save/load?

campagnola commented on neuroanalysis/tests/bath-ugly.h5:
There is a separaate test data repo

outofculture committed:
load data from test_data submodule

outofculture committed:
load wants to be able to construct the stimulus without the entire command channel

outofculture committed:
add bath recording

outofculture committed:
version the test pulse save schema

outofculture committed:
allow TSeries to concat with each other

outofculture committed:
put stimulus in proper format for saving

outofculture committed:
ignore warnings in both places exp_fit is used

outofculture committed:
more warnings it is safe to ignore

outofculture committed:
Refactor PatchClampTestPulse to not be a Recording

- has-a relationship with its recording is better
- refactor how save/load works to understand this better

outofculture committed:
3.10 compat for warnings

outofculture committed:
Recording.save

outofculture committed:
share safety checks between save and copy

outofculture committed:
time vals will now be none to start with
